### PR TITLE
fix(add-demonstrator-as-module)

### PR DESCRIPTION
**Pull request by ba-talibe**
PR for https://github.com/wikit-ai/olaf/issues/94

### CODE REVIEW

**Review by schmarion — CHANGES_REQUESTED** [review]
Missing Readme update with instructions to launch the demo

**Comment by ba-talibe** [comment]
Demo instruction are not yet effective. I added them on next pull request.

**Comment by schmarion** [comment]
If I don't have the instructions, I can't run your code to test it and merge it.

**Comment by ba-talibe** [comment]
So should I add others commits or just give your the commands.

File: README.md [hunk 1]
```diff
@@ -26,6 +26,11 @@ source venv/bin/activate
 pip install .
 ```
 
+For demonstration : 
+
+`olaf run no_llm_pipeline` to run the no llm_pipeline
+`olaf show no_llm_pipeline` to show all components of the no llm_pipeline
+
```
Comment (schmarion) [inline README.md:33]: Put this in the section "Quick-start" with the following text : 

Pipelines can be run with the following command: `olaf run demo_pipeline `.
Pipeline components are displayed with the following command: `olaf show demo_pipeline `
The text can be updated in the file `data/demo.txt`.

=> so I would like you to change the command line from `olaf run no_llm_pipeline` to `olaf turn demo_pipeline` and `olaf show no_llm_pipeline` to `olaf show demo_pipeline`

File: olaf/scripts/no_llm_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        if os.path.isfile(corpus_file):
+            corpus_loader = TextCorpusLoader(corpus_path=corpus_file)
+        elif corpus_path := os.path.isfile(
+            os.path.join(os.getenv("DATA_PATH"), corpus_file)
+        ):
+            corpus_loader = TextCorpusLoader(corpus_path=corpus_path)
+        else:
+            corpus_loader = TextCorpusLoader(
+                corpus_path=os.path.join(os.getenv("DATA_PATH"), "demo.txt")
+            )
```
Comment (schmarion) [inline olaf/scripts/no_llm_pipeline.py:42]: Not sure you need all of these as you already know the file

File: olaf/scripts/no_llm_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        self.pipeline.add_pipeline_component(owl_axiom_extraction)
+
+    def run(self) -> None:
+        """LLM pipeline execution."""
+
+        self.add_pipeline_components()
+        self.pipeline.run()
+
+        kr_serialiser = KRJSONSerialiser()
+        kr_serialisation_path = os.path.join(os.getcwd(), "no_llm_pipeline_kr.json")
```
Comment (schmarion) [inline olaf/scripts/no_llm_pipeline.py:85]: "no_llm_pipeline_kr.json" => "demo_pipeline_kr.json"

File: olaf/scripts/runner.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        elif corpus_path := os.path.isfile(
+            os.path.join(os.getenv("DATA_PATH"), corpus_file)
+        ):
+            corpus_loader = TextCorpusLoader(corpus_path=corpus_path)
+        else:
+            corpus_path = os.path.isfile(
+                os.path.join(os.getenv("DATA_PATH"), "pizza_description.txt")
+            )
+            corpus_loader = TextCorpusLoader(corpus_path=corpus_path)
+        self.pipeline = Pipeline(spacy_model=spacy_model, corpus_loader=corpus_loader)
```
Comment (schmarion) [inline olaf/scripts/runner.py:29]: Redundant with the `no_llm_pipeline.py` init. Either you remove the code from here or in the `no_llm_pipeline.py` init

**Review by schmarion — CHANGES_REQUESTED** [review]
The code is not running for me, either the `run` and `show` commands

File: olaf/scripts/runner.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import os
+import spacy
+from abc import ABC, abstractmethod
+from olaf import Pipeline
+from olaf.repository.corpus_loader import TextCorpusLoader
+
+
+class Runner(ABC):
+    def __init__(self, model_name="en_core_web_md"):
```
Comment (schmarion) [inline olaf/scripts/runner.py:9]: specify the type of `model_name`, not a default value and add `corpus_path` with its type as parameter

File: olaf/scripts/runner.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+        Attributes
+        ----------
+        pipeline: Pipeline
+            The pipeline to execute.
+        """
+        spacy_model = spacy.load(model_name)
+
+        corpus_loader = TextCorpusLoader(
+            corpus_path=os.path.join(os.getenv("DATA_PATH"), "demo.txt")
```
Comment (schmarion) [inline olaf/scripts/runner.py:20]: use the variable `corpus_path` here and precise `corpus_path="demo.txt"` in the init of `demo_pipeline.py`

**Review by schmarion — CHANGES_REQUESTED** [review]
Still not running

File: olaf/scripts/runner.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    def __init__(self, model_name: str, corpus_path: str):
+        """Initialise a pipeline Runner.
+
+        Attributes
+        ----------
+        pipeline: Pipeline
+            The pipeline to execute.
+        """
+        spacy_model = spacy.load(model_name)
+        print(os.getcwd())
```
Comment (schmarion) [inline olaf/scripts/runner.py:18]: remove this

File: olaf/__main__.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    pipelines = []
+    for filename in os.listdir(module_dir):
+        # Check if the file is a Python module and doesn't match the pattern
+        if (
+            filename.endswith(".py")
+            and not re.match(r"(__\w+__|_\w+)\.py", filename)
+            and filename[:-3] != "runner"
+        ):
+            module_name = filename[:-3]  # Remove the ".py" extension
+        pipelines.append(module_name)
```
Comment (schmarion) [inline olaf/__main__.py:32]: what is the point of lines 24 to 32 if line 34 overwrite it ?

Comment (ba-talibe) [reply olaf/__main__.py:32]: I retrieve automatically all files in the `scripts` folder excepted the `runner` file each file is a demo pipeline.
So to create a new pipeline , we have to create a file , in this file create a new subclass of `Runner` class and then implement the `add_pipeline_components` and `run` methods.

File: olaf/scripts/runner.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        self.add_pipeline_components()
+        self.pipeline.run()
+
+        kr_serialiser = KRJSONSerialiser()
+        kr_serialisation_path = os.path.join(os.getcwd(), f"{name}_kr.json")
+        kr_serialiser.serialise(kr=self.pipeline.kr, file_path=kr_serialisation_path)
+
+        kr_rdf_graph_path = os.path.join(
+            os.getcwd(),
+            f"{name}_kr_rdf_graph.ttl",
```
Comment (schmarion) [inline olaf/scripts/runner.py:38]: this will create the results in the scripts folder but it would be better to have it in the data folder

Comment (ba-talibe) [reply olaf/scripts/runner.py:38]: ok why not

**Comment by ba-talibe** [comment]
it's done

File: olaf/__main__.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def run_pipeline(args):
+    print(f"Running pipeline: {args.pipeline} \n")
+
+    try:
+        module = importlib.import_module(f"olaf.scripts.{args.pipeline}")
+    except ModuleNotFoundError:
+        print("Invalid pipeline")
+        sys.exit(1)
```
Comment (schmarion) [inline olaf/__main__.py:53]: after  `print("Invalid pipeline")`, add `list_pipeline()`

File: olaf/__main__.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def show_pipeline(args):
+    print("Showing pipeline:", args.pipeline)
+    try:
+        print(f"olaf.scripts.{args.pipeline}")
+        module = importlib.import_module(f"olaf.scripts.{args.pipeline}")
+    except ModuleNotFoundError:
+        print("Invalid pipeline")
+        sys.exit(1)
```
Comment (schmarion) [inline olaf/__main__.py:70]: after  `print("Invalid pipeline")`, add `list_pipeline()`

**Review by schmarion — APPROVED** [review]

**Pull request merged**